read_utils.py

```diff
@@ -1052,6 +1057,267 @@ def parser_bwamem_idxstats(parser=argparse.ArgumentParser()):
+                        default=1024,
+                        type=int,
+                        help="Width of the plot in pixels (default: %(default)s)")
+    parser.add_argument('--plotHeight',
+                        dest="plot_height",
+                        default=768,
+                        type=int,
+                        help="Width of the plot in pixels (default: %(default)s)")
+    parser.add_argument('--plotTitle',
+                        dest="plot_title",
+                        default="Coverage Plot",
+                        type=str,
+                        help="The title displayed on the coverage plot (default: '%(default)s')")
+    parser.add_argument('-q',
+                        dest="base_q_threshold",
+                        default=None,
+                        type=int,
+                        help="The minimum base quality threshold")
+    parser.add_argument('-Q',
+                        dest="mapping_q_threshold",
+                        default=None,
+                        type=int,
+                        help="The minimum mapping quality threshold")
+    parser.add_argument('-m',
+                        dest="max_coverage_depth",
+                        default=1000000,
+                        type=int,
+                        help="The max coverage depth (default: %(default)s)")
+    parser.add_argument('-l',
+                        dest="read_length_threshold",
+                        default=None,
+                        type=int,
+                        help="Read length threshold")
+    parser.add_argument('--outSummary',
+                        dest="out_summary",
+                        default=None,
+                        type=str,
+                        help="Coverage summary TSV file. Default is to write to temp.")
+    return parser
+
+def plot_coverage(in_bam, out_plot_file, plot_format, plot_style, plot_width, plot_height, plot_title, base_q_threshold, mapping_q_threshold, max_coverage_depth, read_length_threshold, out_summary=None):
+    ''' 
+        Generate a coverage plot from an aligned bam file
+    '''
+    
+    # TODO: remove this:
+    #coverage_tsv_file = "/Users/tomkinsc/Downloads/plottest/test_multisegment.tsv"
+
+    samtools = tools.samtools.SamtoolsTool()
+
+    # check if in_bam is aligned, if not raise an error
+    num_mapped_reads = samtools.count(in_bam, opts=["-F", "4"])
+    if num_mapped_reads == 0:
+        raise Exception("""The bam file specified appears to have zero mapped reads. 'plot_coverage' requires an aligned bam file. You can try 'align_and_plot_coverage' if you don't mind a simple bwa alignment. \n File: %s""" % in_bam)
+
+
+    if out_summary is None:
+        coverage_tsv_file = mkstempfname('.summary.tsv')
+    else:
+        coverage_tsv_file = out_summary
+
+    bam_aligned = mkstempfname('.aligned.bam')
+
+    if in_bam[-4:] == ".sam":
+        # convert sam -> bam
+        samtools.view(["-b"], in_bam, bam_aligned)
```

Comment:
But you always need to sort and index, and since the samtools sort step takes anything as input (and produces bam output), the sam to bam conversion seems unnecessary